### DISCUSSION TITLE
custom .disabled? method

### DISCUSSION THREAD

**Discussion by Valentyna** [ElMassimo/capybara-compose#10] [Q&A] [answered]
When element is disabled in the app under test - it has attribute `[class="k-state-disabled"]`, so regular Capybara `.disabled?` method doesn't work.
I'm new to the TestHelper, love the idea though and would like to learn how to use it.

I want this new method to be available to all Capybara::Node::Elements and I'm trying this:

```
class ElementTestHelper < Capybara::TestHelper
  use_test_helpers(:element)

  def state_disabled?
    has_css?('.k-state-disabled')
  end
end
```
But seeing error:
`NoMethodError: undefined method 'state_disabled?' for #<Capybara::Node::Element:0x00007fee09642d10>`

I'd appreciate if anyone pointed out to me what am I doing wrong

**Comment by ElMassimo** [comment, marked as answer]
Hi Valentyna!

Creating an `ElementTestHelper` will not make the method available in all `Capybara::Node::Element` objects—doing that would require monkeypatching the `Capybara::Node::Element` class.

You would instead need to use the `ElementTestHelper` in a test, as:

```ruby
RSpec.feature 'My Test', test_helpers: [:element] do
  scenario 'checking an element' do
    disabled = element.find('.some-element').state_disabled?
  end
end
```

If you will use this check to perform assertions, it might be preferable to create a [custom assertion](https://capybara-test-helpers.netlify.app/guide/essentials/assertions.html#custom-assertions-%F0%9F%8E%A9) instead of a [check](https://capybara-test-helpers.netlify.app/guide/essentials/querying.html):

```ruby
class ElementTestHelper < Capybara::TestHelper
  def have_state_disabled
    have('.k-state-disabled')
  end
end
```

which you can then use in your tests:

```ruby
button = element.find('.some-button').should_not.have_state_disabled

# For example, if clicking the button should disable it.
button.click.should.have_state_disabled
```

Using assertions ensures the test fails if the condition is not met, and ensures that Capybara retries (preventing flaky tests).

<hr/>

Another (not recommendable) approach would be to define a [custom filter](https://github.com/ElMassimo/capybara-compose/blob/main/spec/support/global_filters.rb#L10-L19) and add it to all [Capybara selectors](https://capybara-test-helpers.netlify.app/api/selectors).

```ruby
# Example: Allow built-in selectors to receive an explicit `:state_disabled` filter option.
#
# find_link(state_disabled: true) # <a class="k-state-disabled">Checkout</a>
# find_link(state_disabled: false) # <a class="some-other-class">Checkout</a>
Capybara::Selector.all.each_value do |selector|
  selector.instance_eval do
    expression_filter(:state_disabled) do |expression, value|
      prefix = value ? '' : '!'
      builder(expression).add_attribute_conditions(class: "#{prefix}k-state-disabled")
    end
    describe_expression_filters do |**options|
      if (value = options[:state_disabled])
        " with state_disabled #{ value.inspect }"
      end
    end
  end
end
```

That would allow you to use `find('.something', state_disabled: true)`, for example, or check `has_selector?(state_disabled: true)`.
